Drop Windows Socket dependency for `randomenv.cpp`

hebasto opened the pull request:
This change drops a dependency on the ws2_32 library for our libbitcoinkernel by switching to [`GetComputerName`](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-getcomputernamew) function.

hebasto commented:
Friendly ping @sipa @TheCharlatan @sipsorcery :)

fanquake commented on src/randomenv.cpp:
```diff
@@ -357,10 +357,21 @@ void RandAddStaticEnv(CSHA512& hasher)
     // Hostname
+#ifdef WIN32
+    // Not using gethostname from Windows Sockets because
+    // the use of the Winsock DLL might be not initiated.
```
>     // the use of the Winsock DLL might be not initiated.

Why not?

hebasto replied:
When using `libbitcoinkernel`, it depends on its user, no?

fanquake replied:
I don't know, given the kernel and it's API/runtime requirements doesn't exist yet. It's also less clear, because there is nothing else in this codebase documenting this requirement, or preventing winsock usage (or any other DLL that might require manual initialization) from inadvertantly ending up back in the kernel.

hebasto replied:
> ... there is nothing else in this codebase ... preventing winsock usage (or any other DLL that might require manual initialization) from inadvertantly ending up back in the kernel.

Well, the upcoming CMake-based build system will likely [complain](https://github.com/hebasto/bitcoin/pull/134) if Winsock dependency is about to be reintroduced.

fanquake replied:
> CMake-based build system will likely complain

Complain about what? Putting aside all the other points here, why are we changing code in our RNG, supposedly based on some thing that CMake may or may not do, that you didn't mention at all in the PR description?

hebasto replied:
Perhaps, I didn't understand your point correctly. Sorry about that.

This PR goal is to remove winsock dependency from libbitcoinkernel. The current build system does not prevent from reintroducing this dependency inadvertantly. But CMake does.

hebasto replied:
> it comes somewhat from vauge / useless comments

They are dropped now.

TheCharlatan replied:
>  then a test or similar should be added that actually enforces that

Could be something as simple as doing a symbol check for `bitcoin-chainstate` and ensuring `WS2_32` is not in its import table. I guess we should be doing this anyway at some point, but not sure where I would put such a check for non-release, experimental binaries.

fanquake replied:
> I guess we should be doing this anyway at some point,

I will try follow up here with something more globally applicable.

fanquake reviewed (commented):
> requires the Winsock DLL being initiated in advance, which is not always guaranteed, for example, in the libbitcoinkernel.

Surely this is the problem that should be getting solved here? This seems like a short-term fixup, and doesn't prevent it from occuring again in the future? Is this something that's a problem elsewhere in the codebase?

Seems similar to #28486, which iirc hasn't been followed up with / fixed properly. If we can't rely on things being available/usable on Windows, seems like we need to rearchitect code that we can do that, rather than continually patching around/catching this ad-hoc.

hebasto commented:
> > requires the Winsock DLL being initiated in advance, which is not always guaranteed, for example, in the libbitcoinkernel.
> 
> Surely this is the problem that should be getting solved here? This seems like a short-term fixup, and doesn't prevent it from occuring again in the future? Is this something that's a problem elsewhere in the codebase?
> 
> Seems similar to #28486, which iirc hasn't been followed up with / fixed properly. If we can't rely on things being available/usable on Windows, seems like we need to rearchitect code that we can do that, rather than continually patching around/catching this ad-hoc.

This PR main goal is to get rid of the ws2_32 library  dependency for the `libbitcoinkernel`.

However, I agree that Window Socket enabling code could be rearchitected.

hebasto commented on src/randomenv.cpp:
```diff
@@ -357,10 +357,21 @@ void RandAddStaticEnv(CSHA512& hasher)
+    char hname[max_size];
+    DWORD size = max_size;
+    if (GetComputerName(hname, &size) != 0) {
+        hasher.Write((const unsigned char*)hname, size);
```
Thanks! Done.

hebasto committed:
Drop Windows Socket dependency for `randomenv.cpp`

This change drops a dependency on the ws2_32 library for our
libbitcoinkernel, which may not be desirable.

sipsorcery commented:
utACK 03b87a3e64305ba651e22a730e35271dea8fea64.

If the hostname hash can be generated without requiring network initialisation that seems like a worthwhile imrpovement.

TheCharlatan commented:
Concept ACK

This does indeed remove the dependency on ws2_32 from the kernel lib.

laanwj commented:
Looks correct to me "If the function succeeds, the return value is a nonzero value.".

I'm not sure the netbios name is always the name as the hostname, but it is a good enough equivalent for randomness purposes, and I agree it is preferable not to depend on winsock for the kernel. The API, introduced in Windows 2000 is also old enough.

Code review ACK 03b87a3e64305ba651e22a730e35271dea8fea64.

fanquake reviewed (approved):
ACK 03b87a3e64305ba651e22a730e35271dea8fea64
